### PR TITLE
Split Stream coordinator into two classes - Client and Service

### PR DESCRIPTION
**Pull request by jguery**
The service part is only running inside Stream HTTP service and does leader election. It also affects a set of Streams to every Stream handler.
The client Stream coordinator is used in the StreamAdmin implementations.

Build is running at: https://builds.cask.co/browse/CDAP-DUT545

### CODE REVIEW

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/InMemoryStreamCoordinatorClient.java [hunk 1]
```diff
@@ -19,35 +19,41 @@
...
 import java.util.Set;
 
 /**
- * In memory implementation for {@link StreamCoordinator}.
+ * In memory implementation for {@link StreamCoordinatorClient}.
  */
 @Singleton
-public final class InMemoryStreamCoordinator extends AbstractStreamCoordinator {
+public final class InMemoryStreamCoordinatorClient extends AbstractStreamCoordinatorClient
+  implements StreamCoordinator {
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/InMemoryStreamCoordinatorClient.java]: Can you align the `implements` and `extends` to make it easier to read?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/DistributedStreamCoordinator.java [hunk 2]
```diff
@@ -76,15 +70,15 @@
   private Cancellable handlerSubscription;
 
   @Inject
-  public DistributedStreamCoordinator(StreamAdmin streamAdmin, ZKClient zkClient,
-                                      DiscoveryServiceClient discoveryServiceClient,
+  public DistributedStreamCoordinator(ZKClient zkClient, DiscoveryServiceClient discoveryServiceClient,
                                       StreamMetaStore streamMetaStore) {
-    super(streamAdmin);
+    super();
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/DistributedStreamCoordinator.java]: No need to call super

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/DistributedStreamCoordinator.java [hunk 3]
```diff
@@ -154,47 +148,48 @@ protected void doShutDown() throws Exception {
   }
 
   @Override
-  protected <T> PropertyStore<T> createPropertyStore(Codec<T> codec) {
-    return ZKPropertyStore.create(zkClient, "/" + Constants.Service.STREAMS + "/properties", codec);
+  public void setHandlerDiscoverable(Discoverable discoverable) {
+
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/DistributedStreamCoordinator.java]: empty method?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/InMemoryStreamCoordinatorClient.java [hunk 4]
```diff
@@ -66,20 +72,57 @@ public void setHandlerDiscoverable(Discoverable discoverable) {
...
     return Futures.immediateFuture(null);
   }
 
-  private void invokeLeaderListeners() throws Exception {
+  private void invokeLeaderListeners(String streamName) throws Exception {
     Set<String> streamNames = Sets.newHashSet();
     for (StreamSpecification spec : streamMetaStore.listStreams()) {
       streamNames.add(spec.getName());
     }
+    if (streamNames.contains(streamName)) {
```
Comment (jguery) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/InMemoryStreamCoordinatorClient.java]: The logic here should be ``if (createdStream != null)``.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/StreamHttpService.java [hunk 5]
```diff
@@ -97,6 +100,12 @@ public InetSocketAddress getSocketAddress() {
     janitorService.startAndWait();
     streamCoordinator.setHandlerDiscoverable(discoverable);
     streamCoordinator.startAndWait();
+
+    // In the in-memory implementations, streamCoordinator and streamCoordinatorLeader are the same implementation,
+    // hence we don't want to start them twice.
+    if (!streamCoordinatorClient.isRunning()) {
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/StreamHttpService.java]: I think it's ok to just call start. Trying to start a already started guava service has no effect.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/StreamHttpService.java [hunk 6]
```diff
@@ -110,6 +119,9 @@ protected void shutDown() throws Exception {
     } finally {
       httpService.stopAndWait();
       streamCoordinator.stopAndWait();
+      if (streamCoordinatorClient.isRunning()) {
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/StreamHttpService.java]: Same here, it's ok to just call stop.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/InMemoryStreamCoordinatorClient.java (hunk 1 above)
Comment (jguery) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/InMemoryStreamCoordinatorClient.java]: Ok.

**Comment by chtyim** [comment]
Just couple comments. No major issue.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/DistributedStreamCoordinator.java (hunk 2 above)
Comment (jguery) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/DistributedStreamCoordinator.java]: Yes, left overs from a previous implementation.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/InMemoryStreamCoordinatorClient.java (hunk 4 above)
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/InMemoryStreamCoordinatorClient.java]: huh? What is `createdStream`?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/DistributedStreamCoordinator.java (hunk 3 above)
Comment (jguery) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/DistributedStreamCoordinator.java]: Will change.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/StreamHttpService.java (hunk 5 above)
Comment (jguery) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/StreamHttpService.java]: Oh, ok then.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/StreamHttpService.java (hunk 6 above)
Comment (jguery) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/StreamHttpService.java]: Ok.

**Comment by jguery** [comment]
Build is green, and there was no major issue. Merging now.

**Pull request merged**